runtime/cudaq/qis/remote_state.cpp

```diff
@@ -0,0 +1,144 @@
+  return state->operator()(tensorIdx, indices);
+}
+
+std::unique_ptr<SimulationState>
+RemoteSimulationState::createFromSizeAndPtr(std::size_t size, void *ptr,
+                                            std::size_t) {
+  throw std::runtime_error("RemoteSimulationState cannot be created from data");
+}
+
+void RemoteSimulationState::dump(std::ostream &os) const {
+  execute();
+  state->dump(os);
+}
+
+cudaq::SimulationState::precision RemoteSimulationState::getPrecision() const {
+  execute();
+  return state->getPrecision();
+}
+
+void RemoteSimulationState::destroyState() { state.reset(); }
+
+std::tuple<std::string, void *, std::size_t>
+RemoteSimulationState::getKernelInfo() const {
+  return std::make_tuple(kernelName, static_cast<void *>(argsBuffer.data()),
+                         argsBuffer.size());
+}
+
+std::vector<std::complex<double>> RemoteSimulationState::getAmplitudes(
+    const std::vector<std::vector<int>> &basisStates) {
+  if (basisStates.empty())
+    return {};
+
+  if (basisStates[0].size() <= maxQubitCountForFullStateTransfer()) {
+    execute();
+    return state->getAmplitudes(basisStates);
+  }
+  auto &platform = cudaq::get_platform();
+  // Create an execution context, indicate this is for
+  // extracting the state representation
+  ExecutionContext context("extract-state");
+  for (const auto &basisState : basisStates)
+    context.amplitudeMaps[basisState] = {};
+  // Perform the usual pattern set the context,
+  // execute and then reset
+  platform.set_exec_ctx(&context);
+  platform.launchKernel(kernelName, nullptr,
+                        static_cast<void *>(argsBuffer.data()),
+                        argsBuffer.size(), 0);
+  platform.reset_exec_ctx();
+  std::vector<std::complex<double>> amplitudes;
+  amplitudes.reserve(basisStates.size());
+  for (const auto &basisState : basisStates)
+    amplitudes.emplace_back(context.amplitudeMaps[basisState]);
+  return amplitudes;
+}
+
+std::complex<double>
+RemoteSimulationState::getAmplitude(const std::vector<int> &basisState) {
+  return getAmplitudes({basisState}).front();
+}
+
+std::complex<double>
+RemoteSimulationState::overlap(const cudaq::SimulationState &other) {
+  const auto &otherState = dynamic_cast<const RemoteSimulationState &>(other);
+  auto &platform = cudaq::get_platform();
+  ExecutionContext context("state-overlap");
+  context.overlapComputeStates =
+      std::make_pair(static_cast<const cudaq::SimulationState *>(this),
+                     static_cast<const cudaq::SimulationState *>(&otherState));
+  platform.set_exec_ctx(&context);
+  platform.launchKernel(kernelName, nullptr, nullptr, 0, 0);
+  platform.reset_exec_ctx();
+  assert(context.overlapResult.has_value());
+  return context.overlapResult.value();
+}
+
+std::size_t RemoteSimulationState::maxQubitCountForFullStateTransfer() {
+  // Default number of qubits for full state vector transfer.
+  constexpr std::size_t NUM_QUBITS_STATE_TRANSFER = 25; // (~100MB of data)
```

Comment:
Since this is server-side code (I think), what happens if the user tries to submit a request exceeding this limit without providing any specific amplitude basis vectors? Do they get an error message or warning message somewhere?

Reply:
It's actually client-side code. The limit here is mainly an optimization for amplitude access (resolving the whole state vector rather than sending amplitude requests). We'll have a doc update to note about the benefit of using batched amplitudes API rather than sequential amplitude calls for large number of qubits using NVQC.

If users want to resolve the whole state vector (not an amplitude access), we treat them as the current v0.7 mode, sending back the whole state vector (memory allocation failures if the state is too big).